// @waves/tx-json-schemas/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemas = exports.schemaTypeMap = exports.validators = void 0;
const Ajv = require("ajv");
const generic_1 = require("@waves/waves-transactions/dist/generic");
const transactions_1 = require("@waves/waves-transactions/dist/transactions");
const schemas_1 = __importDefault(require("./schemas"));
exports.schemas = schemas_1.default;
const ajv = Ajv({
    allErrors: true,
});
exports.validators = generic_1.mapObj(schemas_1.default, (schema) => ajv.compile(schema));
exports.schemaTypeMap = {
    [transactions_1.TRANSACTION_TYPE.ISSUE]: {
        schema: schemas_1.default.IIssueTransaction,
        paramsSchema: schemas_1.default.IIssueParams,
        validator: exports.validators.IIssueTransaction,
        paramsValidator: exports.validators.IIssueParams,
    },
    [transactions_1.TRANSACTION_TYPE.TRANSFER]: {
        schema: schemas_1.default.ITransferTransaction,
        paramsSchema: schemas_1.default.ITransferParams,
        validator: exports.validators.ITransferTransaction,
        paramsValidator: exports.validators.ITransferParams,
    },
    [transactions_1.TRANSACTION_TYPE.REISSUE]: {
        schema: schemas_1.default.IReissueTransaction,
        paramsSchema: schemas_1.default.IReissueParams,
        validator: exports.validators.IReissueTransaction,
        paramsValidator: exports.validators.IReissueParams,
    },
    [transactions_1.TRANSACTION_TYPE.BURN]: {
        schema: schemas_1.default.IBurnTransaction,
        paramsSchema: schemas_1.default.IBurnParams,
        validator: exports.validators.IBurnTransaction,
        paramsValidator: exports.validators.IBurnParams,
    },
    [transactions_1.TRANSACTION_TYPE.EXCHANGE]: {
        schema: schemas_1.default.IExchangeTransaction,
        paramsSchema: schemas_1.default.IExchangeTransaction,
        validator: exports.validators.IExchangeTransaction,
        paramsValidator: exports.validators.IExchangeTransaction,
    },
    [transactions_1.TRANSACTION_TYPE.LEASE]: {
        schema: schemas_1.default.ILeaseTransaction,
        paramsSchema: schemas_1.default.ILeaseParams,
        validator: exports.validators.ILeaseTransaction,
        paramsValidator: exports.validators.ILeaseParams,
    },
    [transactions_1.TRANSACTION_TYPE.CANCEL_LEASE]: {
        schema: schemas_1.default.ICancelLeaseTransaction,
        paramsSchema: schemas_1.default.ICancelLeaseParams,
        validator: exports.validators.ICancelLeaseTransaction,
        paramsValidator: exports.validators.ICancelLeaseParams,
    },
    [transactions_1.TRANSACTION_TYPE.ALIAS]: {
        schema: schemas_1.default.IAliasTransaction,
        paramsSchema: schemas_1.default.IAliasParams,
        validator: exports.validators.IAliasTransaction,
        paramsValidator: exports.validators.IAliasParams,
    },
    [transactions_1.TRANSACTION_TYPE.MASS_TRANSFER]: {
        schema: schemas_1.default.IMassTransferTransaction,
        paramsSchema: schemas_1.default.IMassTransferParams,
        validator: exports.validators.IMassTransferTransaction,
        paramsValidator: exports.validators.IMassTransferParams,
    },
    [transactions_1.TRANSACTION_TYPE.DATA]: {
        schema: schemas_1.default.IDataTransaction,
        paramsSchema: schemas_1.default.IDataParams,
        validator: exports.validators.IDataTransaction,
        paramsValidator: exports.validators.IDataParams,
    },
    [transactions_1.TRANSACTION_TYPE.SET_SCRIPT]: {
        schema: schemas_1.default.ISetScriptTransaction,
        paramsSchema: schemas_1.default.ISetScriptParams,
        validator: exports.validators.ISetScriptTransaction,
        paramsValidator: exports.validators.ISetScriptParams,
    },
    [transactions_1.TRANSACTION_TYPE.SPONSORSHIP]: {
        schema: schemas_1.default.ISponsorshipTransaction,
        paramsSchema: schemas_1.default.ISponsorshipParams,
        validator: exports.validators.ISponsorshipTransaction,
        paramsValidator: exports.validators.ISponsorshipParams,
    },
    [transactions_1.TRANSACTION_TYPE.SET_ASSET_SCRIPT]: {
        schema: schemas_1.default.ISetAssetScriptTransaction,
        paramsSchema: schemas_1.default.ISetAssetScriptParams,
        validator: exports.validators.ISetAssetScriptTransaction,
        paramsValidator: exports.validators.ISetAssetScriptParams,
    },
    [transactions_1.TRANSACTION_TYPE.INVOKE_SCRIPT]: {
        schema: schemas_1.default.IInvokeScriptTransaction,
        paramsSchema: schemas_1.default.IInvokeScriptParams,
        validator: exports.validators.IInvokeScriptTransaction,
        paramsValidator: exports.validators.IInvokeScriptParams,
    },
    [transactions_1.TRANSACTION_TYPE.UPDATE_ASSET_INFO]: {
        schema: schemas_1.default.IUpdateAssetInfoTransaction,
        paramsSchema: schemas_1.default.IUpdateAssetInfoParams,
        validator: exports.validators.IUpdateAssetInfoTransaction,
        paramsValidator: exports.validators.IUpdateAssetInfoParams,
    }
};
