// @waves/node-api-js/es/api-node/addresses/index.js
import request from '../../tools/request';
import query from '../../tools/query';
export function fetchDataKey(base, address, key, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/data/${address}/${encodeURIComponent(key)}`,
        options
    });
}
export function fetchScriptInfoMeta(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/scriptInfo/${address}/meta`,
        options
    });
}
export function fetchBalanceDetails(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/details/${address}`,
        options
    });
}
export function fetchBalanceConfirmations(base, address, confirmations, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/${address}/${confirmations}`,
        options
    });
}
export function fetchScriptInfo(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/scriptInfo/${address}`,
        options
    });
}
export function data(base, address, params = Object.create(null), options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/data/${address}${query(params)}`,
        options
    });
}
export function fetchValidate(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/validate/${address}`,
        options
    });
}
export function fetchBalance(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/${address}`,
        options
    });
}
export function fetchEffectiveBalanceConfirmations(base, address, confirmations, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/effectiveBalance/${address}/${confirmations}`,
        options
    });
}
export function fetchEffectiveBalance(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/effectiveBalance/${address}`,
        options
    });
}
export function fetchSeq(base, from, to, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/seq/${from}/${to}`,
        options
    });
}
export function fetchSeed(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/seed/${address}`,
        options
    });
}
export function fetchPublicKey(base, publicKey, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/publicKey/${publicKey}`,
        options
    });
}
export function fetchAddresses(base, options = Object.create(null)) {
    return request({
        base,
        url: '/addresses',
        options
    });
}
