// @waves/node-api-js/es/api-node/addresses/index.js
import request, { parseResponse } from '../../tools/request';
import query from '../../tools/query';
export function fetchDataKey(base, address, key, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/data/${address}/${encodeURIComponent(key)}`,
        options
    });
}
export function fetchScriptInfo(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/scriptInfo/${address}`,
        options
    });
}
export function fetchScriptInfoMeta(base, address) {
    return request({
        base,
        url: `/addresses/scriptInfo/${address}/meta`
    });
}
export function fetchBalanceDetails(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/details/${address}`,
        options
    });
}
export function fetchBalanceConfirmations(base, address, confirmations, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/${address}/${confirmations}`,
        options
    });
}
export function data(base, address, params = Object.create(null), options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/data/${address}${query(params)}`,
        options
    });
}
export function fetchValidate(base, address) {
    return request({
        base,
        url: `/addresses/validate/${address}`
    });
}
export function fetchBalance(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/balance/${address}`,
        options
    });
}
export function fetchMultipleBalance(base, addresses, options = Object.create(null)) {
    return fetch(`${base}/addresses/balance`, {
        method: "POST",
        body: JSON.stringify({ addresses }),
        headers: {
            "Content-Type": "application/json"
        }
    }).then(parseResponse);
}
export function deleteAddressFromWallet(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/${address}`,
        options: Object.assign(Object.assign({}, options), { method: 'DELETE' })
    });
}
export function fetchEffectiveBalanceConfirmations(base, address, confirmations, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/effectiveBalance/${address}/${confirmations}`,
        options
    });
}
export function fetchEffectiveBalance(base, address, options = Object.create(null)) {
    return request({
        base,
        url: `/addresses/effectiveBalance/${address}`,
        options
    });
}
export function fetchSeq(base, from, to) {
    return request({
        base,
        url: `/addresses/seq/${from}/${to}`
    });
}
export function fetchSeed(base, address) {
    return request({
        base,
        url: `/addresses/seed/${address}`
    });
}
export function fetchPublicKey(base, publicKey) {
    return request({
        base,
        url: `/addresses/publicKey/${publicKey}`
    });
}
export function fetchAddresses(base) {
    return request({
        base,
        url: '/addresses'
    });
}
